### PR TITLE
bash: upgrade default version to 5

### PR DESCRIPTION
**Pull request by happysalada**
###### Motivation for this change

bash default version upgrade

cc @SuperSandro2000 @06kellyjac 

@vcunat just leaving a comment here to test this on hydra at some point to verify nothing major breaks. (no rush, I know you are busy).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
related: #131681

please update the PR title to match contributing, the same as the last one would do.

Also FYI you can make a new separate branch and push it over the one you were using for the last PR
`git push <remote> <branch with new changes>:<branch you are pushing to> `

e.g. `git push origin upgrade_bash:bash_default_upgrade --force`

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10507,20 +10507,20 @@ with pkgs;
...
     binutils = stdenv.cc.bintools;
   });
-  bashInteractive_5 = lowPrio (callPackage ../shells/bash/5.1.nix {
+  bashInteractive = lowPrio (callPackage ../shells/bash/5.1.nix {
     binutils = stdenv.cc.bintools;
     interactive = true;
     withDocs = true;
   });
 
   # WARNING: this attribute is used by nix-shell so it shouldn't be removed/renamed
```
Comment (06kellyjac) [inline pkgs/top-level/all-packages.nix:10522]: This comment needs to be moved above `bashInteractive = lowPrio (callPackage ../shells/bash/5.1.nix {` now

Comment (happysalada) [reply pkgs/top-level/all-packages.nix:10522]: thank you

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -10507,20 +10507,20 @@ with pkgs;
...
+  bash_4 = lowPrio (callPackage ../shells/bash/4.4.nix {
     binutils = stdenv.cc.bintools;
   });
-  bash_5 = lowPrio (callPackage ../shells/bash/5.1.nix {
+  bash = lowPrio (callPackage ../shells/bash/5.1.nix {
     binutils = stdenv.cc.bintools;
   });
-  bashInteractive_5 = lowPrio (callPackage ../shells/bash/5.1.nix {
+  # WARNING: this attribute is used by nix-shell so it shouldn't be removed/renamed
+  bashInteractive = lowPrio (callPackage ../shells/bash/5.1.nix {
```
Comment (andersk) [inline pkgs/top-level/all-packages.nix:10517]: This `lowPrio` should be moved from `bashInteractive` (5) to `bashInteractive_4`.

Comment (happysalada) [reply pkgs/top-level/all-packages.nix:10517]: Well spotted, thank you!

**Comment by happysalada** [comment]
@vcunat when you have a moment, if you can run a hydra jobset for this PR, it would be great. I know you are busy, no rush.
This PR is fairly simple, so hopefully it should go smoothly.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (happysalada) [reply pkgs/top-level/all-packages.nix:10517]: fixed.

**Comment by andersk** [comment]
Should we keep `bash_5` and `bashInteractive_5` around as aliases (or vice versa) for those who might have already been using them?

**Comment by happysalada** [comment]
I personally would tend to say no, as the upgrade should be straightforward. (I know I'm biased toward breaking things, to signify that something has changed, so I'm open to other opinions).
Perhaps we should add something in the next release notes ? I would prefer to do that in another PR after this one has been merged.

**Comment by 06kellyjac** [comment]
I don't see a reason not to do an update to the release notes and docs in this same PR

**Comment by happysalada** [comment]
Fair enough. I've just added one line to the release notes (I was modelling what the postgres update says basically). I'm happy to change/add more if anybody feels it's needed.
Since this PR might be open for sometime, if a conflict arise, I plan to have the hydra test and do a rebase and merge after.

**Comment by happysalada** [comment]
Hmmm there seems to be a formatting check failure. However the error is just exit 1. If anybody knows why that check is failing, i'm interested.

**Comment by andersk** [comment]
The failing check is `.github/workflows/nixos-manual.yml`. It’s checking that the XML file has been updated from the Markdown file, which it hasn’t. You need to run `nixos/doc/manual/md-to-db.sh`.

**Comment by SuperSandro2000** [comment]
> Should we keep `bash_5` and `bashInteractive_5` around as aliases (or vice versa) for those who might have already been using them?

Yes, please but they can be in aliases.nix.

**Comment by happysalada** [comment]
I've added aliases, and fixed the docs. Let me know if this needs more love.

File: pkgs/top-level/aliases.nix [hunk 3]
```diff
@@ -66,6 +66,8 @@ mapAliases ({
   badtouch = authoscope; # Project was renamed, added 20210626
   bar-xft = lemonbar-xft;  # added 2015-01-16
   bashCompletion = bash-completion; # Added 2016-09-28
+  bash_5 = bash;
+  bashInteractive_5 = bashInteractive;
```
Comment (SuperSandro2000) [inline pkgs/top-level/aliases.nix:70]:
```suggestion
  bash_5 = bash; # added 2021-08-20
  bashInteractive_5 = bashInteractive; # added 2021-08-20
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
The last commit does not follow the contributing guide.

**Comment by happysalada** [comment]
Took me a bit to squash, I was hoping a little for a hydra test this weekend.
I'm still not ready to do those by myself.

**Comment by happysalada** [comment]
@vcunat if you have some time this weekend to start a hydra job for this PR, it would be great. No worries at all if you are busy.

**Comment by vcunat** [comment]
Hydra is quite busy with security rebuild from OpenSSL (master now, 21.05 soon), so it will take at least a few days before there's capacity for that.

**Comment by happysalada** [comment]
No worries, I'll ping again next week. Thank you for having checked!

**Comment by happysalada** [comment]
A hydra job has been created on aarch64-linux for this https://hydra.nixos.org/jobset/nixpkgs/bash-5#tabs-evaluations
no failure due to this upgrade has been detected.
It's not a complete assurance that things are working, but it's a strong enough signal to merge in my opinion.
I will wait a couple more days in case people are against this.

**Comment by nrdxp** [comment]
I would actually merge this right now, but it looks like some conflicts have cropped up...

**Comment by happysalada** [comment]
Just rebased, merging then. Thank you all for the time and care given to this!

**Comment by trofi** [comment]
Broke one `mrsh` test: https://hydra.nixos.org/build/155384906 (`bash` changed behaviour).

**Comment by happysalada** [comment]
@trofi sorry about that! The failure isn't super clear from reading the logs. I'm curious about the change of behavior if you have more information. You probably know this but the bash changelog is not complete and so it makes it very hard to know what will actually break.

**Comment by trofi** [comment]
I only superficially looked at the test. It compares the output from `mrsh` (unchanged) to ouput from `bash` on the same test script (bash changed from 4 to 5). Let me try to dig the details out.

**Comment by trofi** [comment]
The following command changed behaviour under `sh` (but not bash):

```
a=~/foo:~/bar:~/baz; echo $a
```

```
$ nix build -f '<nixpkgs>' bash bash_4
$ ls -l
lrwxrwxrwx 1 slyfox users 55 Nov  1 21:52 result -> /nix/store/wadmyilr414n7bimxysbny876i2vlm5r-bash-5.1-p8
lrwxrwxrwx 1 slyfox users 56 Nov  1 21:52 result-1 -> /nix/store/4y0aacvxrkj3v7wayxpfxjda3km7c72a-bash-4.4-p23

$ /nix/store/wadmyilr414n7bimxysbny876i2vlm5r-bash-5.1-p8/bin/sh -c 'a=~/foo:~/bar:~/baz; echo $a'
/home/slyfox/foo:~/bar:~/baz
$ /nix/store/wadmyilr414n7bimxysbny876i2vlm5r-bash-5.1-p8/bin/bash -c 'a=~/foo:~/bar:~/baz; echo $a'
/home/slyfox/foo:/home/slyfox/bar:/home/slyfox/baz
$ /nix/store/4y0aacvxrkj3v7wayxpfxjda3km7c72a-bash-4.4-p23/bin/bash -c 'a=~/foo:~/bar:~/baz; echo $a'
/home/slyfox/foo:/home/slyfox/bar:/home/slyfox/baz
$ /nix/store/4y0aacvxrkj3v7wayxpfxjda3km7c72a-bash-4.4-p23/bin/sh -c 'a=~/foo:~/bar:~/baz; echo $a'
/home/slyfox/foo:/home/slyfox/bar:/home/slyfox/baz
```

`sh` under `bash-5` stopped expanding `~` in the middle of a string before `:`. Not sure if it's intended or not.

**Comment by andersk** [comment]
Looks intentional. From the bash-5.1-beta changelog:

> 1. Changes to Bash
>    …
>    g. Avoid performing tilde expansion after `:' in words that look like assignment statements when in posix mode.

**Comment by andersk** [comment]
I reported a Bash bug about the apparent POSIX violation with some more context. I’ll link it here when it shows up in the archive.

**Pull request merged**